pkg/gatewayserver/io/downlink_token.go

```diff
@@ -57,7 +57,7 @@ func (t *DownlinkTokens) Next(msg *ttnpb.DownlinkMessage, time time.Time) uint16
 func (t DownlinkTokens) Get(token uint16, time time.Time) (*ttnpb.DownlinkMessage, time.Duration, bool) {
 	pos := token % downlinkTokenItems
 	item := t.items[pos]
-	if item.key != token {
+	if item.key != token || item.msg == nil {
```

Comment:
This is fine, but you may also check if the time is 0.

Reply:
Indeed. I don't see a case where that should be possible but better to do that here.